source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -53,16 +55,34 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   }
   cluster_info_ = cluster->info();
 
-  auto& resource = cluster_info_->resourceManager(route_entry->priority()).pendingRequests();
-  if (!resource.canCreate()) {
+  // By default, envoy doesn't allow to use dns cache circuit breakers even if configured. But this
+  // runtime feature will enable you to use this feature. This runtime feature is prepared to
+  // guarantee backward compatibility.
+  const auto& dns_cache_resource_manager = config_->cache().dnsCacheResourceManager();
+  const bool should_use_dns_cache_circuit_breakers =
+      Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.enable_dns_cache_circuit_breakers") &&
+      dns_cache_resource_manager != nullptr;
+  ResourceLimit& pending_requests =
+      should_use_dns_cache_circuit_breakers
+          ? dns_cache_resource_manager->pendingRequests()
+          : cluster_info_->resourceManager(route_entry->priority()).pendingRequests();
+
+  if (!pending_requests.canCreate()) {
+    // When we use DNS Cache manager circuit breakers, we don't have to take care of upstream
+    // pending counter. Because DNSCacheCircuitBreakersStats will do that by counting
+    // rq_pending_remaining_.
```

Comment:
Where is the overflow counter when using the DNS cache circuit breaker? I think you still need one?

Reply:
Added overflow counter to dns cache stats.